src/test/java/com/labs/introtoprogramming/lab4/image/io/bmp/BMPImageWriterTests.java

```diff
@@ -1,16 +1,145 @@
+    writer.close();
+  }
+
+  @Test
+  public void testPixelStorageRowByteBuffer() throws IOException {
+    BMPImageWriter writer = new BMPImageWriter(new ByteArrayOutputStream());
+    assertArrayEquals(
+            new byte[] { 0, 0, (byte) 0xFF, 0, (byte) 0xBB, 0, 0, 0 },
+            writer.pixelStorageRowAsByteBuffer(DUMMY_IMAGE, 0).array()
+    );
+    writer.close();
+  }
+
+  @Test
+  public void testPixelStorageByteBuffer() throws IOException {
+    BMPImageWriter writer = new BMPImageWriter(new ByteArrayOutputStream());
+    assertArrayEquals(
+            new byte[] {
+                    0, 0, (byte) 0xFF, 0,
+                    (byte) 0xBB, 0, 0, 0,
+                    (byte) 0xDD, 0, 0, (byte) 0xEE,
+                    (byte) 0xCC, (byte) 0xAA, 0, 0
+            },
+            writer.pixelStorageAsByteBuffer(DUMMY_IMAGE).array()
+    );
+    writer.close();
+  }
+
+  @Test
+  public void testDIBHeaderByteBuffer() throws IOException {
+    BMPImageWriter writer = new BMPImageWriter(new ByteArrayOutputStream());
+    assertArrayEquals(
+            new byte[] {
+                    40, 0, 0, 0, 2, 0, 0, 0, 2, 0,
+                    0, 0, 1, 0, 24, 0, 0, 0, 0, 0,
+                    16, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+            },
+            writer.dibHeaderAsByteBuffer(DUMMY_IMAGE, 16).array()
+    );
+    writer.close();
+  }
+
+  @Test
+  public void testHeaderByteBuffer() throws IOException {
+    BMPImageWriter writer = new BMPImageWriter(new ByteArrayOutputStream());
+    assertArrayEquals(
+            new byte[] {
+                    0x42, 0x4D,
+                    14 + 40 + 16, 0, 0, 0,
+                    0, 0, 0, 0,
+                    14 + 40, 0, 0, 0
+            },
+            writer.headerAsByteBuffer(40, 16).array()
+    );
+    writer.close();
+  }
+
   @Test
-  public void testImageWrite() {
+  public void testWritingImage() throws IOException {
     ByteArrayOutputStream output = new ByteArrayOutputStream();
     BMPImageWriter writer = new BMPImageWriter(output);
-    writer.write(new RGBImage(0, 0, null, null, null));
+    writer.write(DUMMY_IMAGE);
+    writer.close();
+
+    ImageReader reader = ImageIO.getImageReadersByFormatName("bmp").next();
+    ImageInputStream imageInputStream = ImageIO.createImageInputStream(new ByteArrayInputStream(output.toByteArray()));
+    reader.setInput(imageInputStream, true);
+
+    BufferedImage image = reader.read(0);
+    assertEquals(new Color(255, 0, 0).getRGB(), image.getRGB(0, 1));
+    assertEquals(new Color(0, 187, 0).getRGB(), image.getRGB(1, 1));
+    assertEquals(new Color(0, 0, 221).getRGB(), image.getRGB(0, 0));
+    assertEquals(new Color(170, 204, 238).getRGB(), image.getRGB(1, 0));
+  }
+
+  @Test
+  public void testWriteAndReadImage() throws IOException, UnsupportedDataFormatException {
+    /*ByteArrayOutputStream output = new ByteArrayOutputStream();
```

Comment:
Uncomment after [https://github.com/nikitavbv/ProgrammingAssignment4/pull/9](url) is aproved